File: smalivm/src/main/java/org/cf/smalivm/opcode/MoveOp.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.context.ExecutionNode;
import org.cf.smalivm.context.HeapItem;
import org.cf.smalivm.context.MethodState;
import org.jf.dexlib2.builder.MethodLocation;

public class MoveOp extends MethodStateOp {

    private final MoveType moveType;
    private final int toRegister;
    private int targetRegister;

    MoveOp(MethodLocation location, MethodLocation child, int toRegister, int targetRegister) {
        this(location, child, toRegister, MoveType.REGISTER);
        this.targetRegister = targetRegister;
    }

    MoveOp(MethodLocation location, MethodLocation child, int toRegister, MoveType moveType) {
        super(location, child);
        this.toRegister = toRegister;
        this.moveType = moveType;
    }

    private static void moveException(MethodState mState, int toRegister) {
        HeapItem exception = mState.peekExceptionRegister();
        mState.assignRegister(toRegister, exception);
    }

    private static void moveRegister(MethodState mState, int toRegister, int fromRegister) {
        HeapItem item = mState.readRegister(fromRegister);
        mState.assignRegister(toRegister, item);
    }

    private static void moveResult(MethodState mState, int toRegister) {
        HeapItem item = mState.readResultRegister();
        mState.assignRegister(toRegister, item);
    }

    @Override
    public void execute(ExecutionNode node, MethodState mState) {
        switch (moveType) {
            case EXCEPTION:
                moveException(mState, toRegister);
                break;
            case RESULT:
                moveResult(mState, toRegister);
                break;
            case REGISTER:
                moveRegister(mState, toRegister, targetRegister);
                break;
        }
    }

    public MoveType getMoveType() {
        return moveType;
    }

    public int getToRegister() {
        return toRegister;
    }

    public int getTargetRegister() {
        return targetRegister;
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder(getName());
        sb.append(" r").append(toRegister);
        if (MoveType.REGISTER == moveType) {
            sb.append(", r").append(targetRegister);
        }

        return sb.toString();
    }

    enum MoveType {
        EXCEPTION, REGISTER, RESULT
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/MoveOp.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.configuration.Configuration
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.type.ClassManager
import org.cf.smalivm2.ExecutionNode
import org.cf.smalivm2.ExecutionState
import org.cf.smalivm2.UnresolvedChild
import org.cf.util.Utils
import org.jf.dexlib2.builder.BuilderInstruction
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.OneRegisterInstruction
import org.jf.dexlib2.iface.instruction.TwoRegisterInstruction

class MoveOp internal constructor(location: MethodLocation, child: MethodLocation, val toRegister: Int, val moveType: MoveType) :
    Op(location, child) {
    var targetRegister = 0
        private set

    internal constructor(location: MethodLocation, child: MethodLocation, toRegister: Int, targetRegister: Int) : this(
        location,
        child,
        toRegister,
        MoveType.REGISTER
    ) {
        this.targetRegister = targetRegister
    }

    override val registersReadCount = 1
    override val registersAssignedCount = 1

    override fun execute(node: ExecutionNode): Array<out UnresolvedChild> {
        when (moveType) {
            MoveType.EXCEPTION -> moveException(node.state, toRegister)
            MoveType.RESULT -> moveResult(node.state, toRegister)
            MoveType.REGISTER -> moveRegister(node.state, toRegister, targetRegister)
        }
        return finishOp()
    }

    override fun toString(): String {
        val sb = StringBuilder(name)
        sb.append(" r").append(toRegister)
        if (MoveType.REGISTER == moveType) {
            sb.append(", r").append(targetRegister)
        }
        return sb.toString()
    }

    enum class MoveType {
        EXCEPTION, REGISTER, RESULT
    }

    companion object : OpFactory {
        private fun getMoveType(opName: String): MoveType {
            return when {
                opName.contains("-result") -> MoveType.RESULT
                opName.contains("-exception") -> MoveType.EXCEPTION
                else -> MoveType.REGISTER
            }
        }

        private fun moveException(state: ExecutionState, toRegister: Int) {
            val exception = state.peekExceptionRegister()!!
            state.assignRegister(toRegister, exception)
        }

        private fun moveRegister(state: ExecutionState, toRegister: Int, fromRegister: Int) {
            val value = state.readRegister(fromRegister)
            state.assignRegister(toRegister, value)
        }

        private fun moveResult(state: ExecutionState, toRegister: Int) {
            val value = state.readResultRegister()
            state.assignRegister(toRegister, value)
        }

        override fun build(
            location: MethodLocation,
            addressToLocation: Map<Int, MethodLocation>,
            classManager: ClassManager,
            classLoader: SmaliClassLoader,
            configuration: Configuration
        ): Op {
            val child = Utils.getNextLocation(location, addressToLocation)
            val instruction = location.instruction as BuilderInstruction
            val opName = instruction.opcode.name
            val toRegister = (instruction as OneRegisterInstruction).registerA
            return when (val moveType = getMoveType(opName)) {
                MoveType.RESULT, MoveType.EXCEPTION -> MoveOp(location, child, toRegister, moveType)
                MoveType.REGISTER -> {
                    val targetRegister = (instruction as TwoRegisterInstruction).registerB
                    MoveOp(location, child, toRegister, targetRegister)
                }
            }
        }
    }
}
